chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,418 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! The ChaCha20 stream cipher from RFC8439.
+
+/// The first four words (32-bit) of the ChaCha stream cipher state are constants.
+const WORD_1: u32 = 0x61707865;
+const WORD_2: u32 = 0x3320646e;
+const WORD_3: u32 = 0x79622d32;
+const WORD_4: u32 = 0x6b206574;
+
+/// The quarter round indexes of a ChaCha block function.
+const CHACHA_ROUND_INDICIES: [(usize, usize, usize, usize); 8] = [
+    // The first 4 rounds are columns of the 4x4 matrix.
+    (0, 4, 8, 12),
+    (1, 5, 9, 13),
+    (2, 6, 10, 14),
+    (3, 7, 11, 15),
+    // The next 4 rounds are diagonals of the 4x4 matrix.
+    (0, 5, 10, 15),
+    (1, 6, 11, 12),
+    (2, 7, 8, 13),
+    (3, 4, 9, 14),
+];
+
+/// The cipher's block size is 64 bytes.
+const CHACHA_BLOCKSIZE: usize = 64;
+
+/// A 256-bit secret session key shared by the parties communicating.
+#[derive(Clone, Copy, Debug)]
+pub struct SessionKey([u8; 32]);
+
+impl SessionKey {
+    /// Create a new session key.
+    pub fn new(key: [u8; 32]) -> Self {
+        SessionKey(key)
+    }
+}
+
+/// A 96-bit initialization vector (IV), or nonce.
+#[derive(Clone, Copy, Debug)]
+pub struct Nonce([u8; 12]);
+
+impl Nonce {
+    /// Create a new nonce.
+    pub fn new(nonce: [u8; 12]) -> Self {
+        Nonce(nonce)
+    }
+}
+
+/// The 512-bit cipher state is chunk'd up into 16 32-bit words.
+///
+/// The 16 words can be visualized as a 4x4 matrix:
+///
+///   0   1   2   3
+///   4   5   6   7
+///   8   9  10  11
+///   12  13  14  15
+#[derive(Clone, Copy, Debug)]
+struct State([u32; 16]);
```

Comment:
I crunched down the other feedback into a single commit following Tobin's advice, so happy to hold it here for the SIMD stuff. I'll take a stab at porting over the LDK SIMD type to learn what is going on there, but won't be offended if we wanna just roll with the LDK implementation in the future too.

Reply:
Is this comment thread asking for changes that should block the PR being merged?

Reply:
The `State` type at least is private, as is its inner field, so I think we can replace it with something more SIMD-compatible even after release.

Reply:
Fwiw I think I can have the SIMD stuff wrapped up today, starting to click for me.

Reply:
I have a SIMD-friendly version here which is generating almost pure SIMD instructions for the quarter round from what I can tell. Took me a few iterations to get the data structures to play nice with the compiler, and I want to test to see how many of the changes are actually required (like I am not sure yet if the State really needs to have separate fields). But otherwise I think it looks pretty clean and it didn't require a full re-write of anything. I have the `U32x4` type matching the current experimental version in the Rust std lib.

I find the vectorize'd version much harder to read than the simple scalar one, so going to be working on a lot of documentation for it. But let me know if anyone has hesitations to just roll this into this PR: https://gist.github.com/nyonson/33660ca17d9062a130dcaadc05278de1